### ISSUE TITLE
Keep repository set in sync with config

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#2025]
### User story

As a site admin, I configure Sourcegraph with a personal access token. I want to share Sourcegraph with my team, but I don't want Sourcegraph to be configured with all of the repositories that my token has access to. Either I update the token, or I limit the scope of synced repositories (i.e. via a search query or regex) and the repositories that are no longer accessible immediately disappear from my Sourcegraph instance.

### Product changes

- Mirrored repositories will be always kept in sync with associated external service configuration.
- Mirrored repositories will be configured **exclusively** via associated external service configuration.
  - Enabled and disabled state will be deprecated and removed from the UI.
  - Explicit deletion of individual repositories will be removed from the UI.

### Planned work for *3.3*

- [x] **Deployment**
   - [x] Instrument Syncer with Prometheus metrics (#3398)
   - [x] Instrument server endpoints with Prometheus metrics (#3398)
   - [x] Instrument new code with tracing. (#3398)
- [x] **Documentation**
   - [x] Update GitHub external service documentation.
      - [x] Document new `repositoryQuery` default. (#2895)
      - [x] Document new `exclude` setting.
   - [x] Update GitLab external service documentation (#3042)
   - [x] Update Bitbucket Server external service documentation.
       - [x] Document new `repositoryQuery` setting.
       - [x] Document new `repos` setting.
       - [x] Document new `exclude` setting.
       - [x] Document new required `username` field.
   - [x] Write blog post about how the new syncer works.
- [x] **User Interface**
   - [x] Implement [new settings page](https://app.zeplin.io/project/5c9bdd56361d0e05516ab7a6/screen/5c9be972ff87da691c574663) in development by @francisschmaltz.
    - [x] Change repos listing page to support repos with and without enabled state and delete buttons.
    - Sourcegraph Engineer / Design Spec Link: https://zpl.io/anpXvGr
    - Public / Non-Engineer Design Link:  https://sketch.cloud/s/wwoGV/Axae1bR/play
serving requests.
- [x] **Common**
    - [x] Ensure enabled state migration is safe to run many times (#3046)
    - [x] Retry migrations instead of exiting. Verify that Sourcegraph is usable without repo-updater (#3046)
    - [x] Trigger full sync when an external service is deleted. (#3290)
    - [x] Ensure that database is migrated once the frontend is ready. Right now waiting for the frontend doesn't ensure that migrations were ran. We either change the wait for frontend method to wait for migrations too, or make repo-updater also run migrations. (#3286)
    - [x] GraphQL API's setRepositoryEnabled deprecation path (#3162, #3252)
    - [x] Handle repo renames in the scheduler (#2609)
    - [x] Default feature flag to on.
    - [x] Use centralized service discovery mechanism to retrieve Postgres connection data. (cc: @slimsag)
- [x] **GitHub**
   - [x] Implement GitHubSource and enable new syncer for GitHub external services (#2266)
   - [x] Implement `github.exclude` to support enabled state deprecation. (#2649)
   - [x] Change default `github.repositoryQuery` to `["none"]` and migrate existing configs (#2785, #2792)
   - [x] Implement enabled state deprecation migration of GitHub external services (#2887)
      - [x] Migrate all disabled repos to `github.exclude` list.
      - [x] Migrate all enabled repos that aren't yielded by any GitHub source to `github.repos` list.
      - [x] Deprecate `github.initialRepositoryEnablement` setting.
   - [x] Show error to admin for repositoryQuery with more than 1000 results (#2562)
   - [x] repositoryQuery may return incomplete results (#3363)
   - [x] Gracefully handle non-existent repos in repos list. (#3027)
   - [x] Update UI to remove enabled state for GitHub repos.
   - [x] Update UI to remove explicit deletion for GitHub repos.
- [x] **GitLab**
   - [x] Implement GitLabSource and enable new syncer for GitLab external services (#3008)
   - [x] Implement `gitlab.projects` to support enabled state deprecation. (#3027)
      - Gracefully handle non-existent projects in projects list.
   - [x] Implement `gitlab.exclude` to support enabled state deprecation. (#3014)
   - [x] Change default `gitlab.projectQuery` to `["none"]` and migrate existing configs. (#3028)
   - [x] Extend enabled state deprecation migration with GitLab external services. (#3029)
   - [x] Update UI to remove enabled state for GitLab repos.
   - [x] Update UI to remove explicit deletion for GitLab repos.

### Stretch goals for *3.3*

- [ ] Implement dry-run mode in `Syncer.Sync` to support #2866
- [x] **Bitbucket Server** 
   - [x] Implement BitbucketServerSource and enable new syncer for BitbucketServer external services (#3097)
   - [x] Implement `bitbucketserver.repos` to support enabled state deprecation. (#3097)
      - Gracefully handle non-existent repos in repos list.
   - [x] Implement `bitbucketserver.exclude` to support enabled state deprecation. (#3097)
   - [x] Implement `bitbucketserver.repositoryQuery`, set the default to `["none"]` and migrate existing configs to have the previous semantically equivalent default of `["?visibility=private", "?visibility=public"]` (#3097)
   - [x] Migrate existing Bitbucket Server repos external_id's from `projectName/repoName` to `repoID` to support renames. (#3097)
   - [x] Implement enabled state deprecation migration of BitbucketServer external services (#3097)
      - [x] Migrate all disabled repos to `bitbucketserver.exclude` list.
      - [x] Migrate all enabled repos that aren't yielded by any BitbucketServer source to `bitbucketserver.repos` list.
      - [x] Deprecate `bitbucketserver.initialRepositoryEnablement` setting.
   - [x] Update UI to remove enabled state for BitbucketServer repos.
   - [x] Update UI to remove explicit deletion for BitbucketServer repos.

### Backlog
- [ ] Create Grafana dashboard for repo-updater
- [ ] Ensure that sourcegraph.com code paths work with the feature flag on. How does manually adding a repo when directly visiting one interact with the new Syncer?
- [ ] Search and replace everywhere: s/Github/GitHub/g
- [ ] Test: Server HTTP endpoints
- [ ] Test: Update from diff in scheduler
- [ ] GitHub supports `query` configuration to replace `repositoryQuery` #2562

### Customers affected

Too many to name individually, but a highlight:

- https://app.hubspot.com/contacts/2762526/company/732945850 said 90% of their site admin's time was managing repositories

**Comment by keegancsmith** [comment]
Something to keep in mind if we make the primitive "repoQuery" is that github's search API has a rate limit of 20 requests an hour. This will be fine when reacting to config changes, but may back us into a corner when we want to notice changes in the search results. Additionally a search request only returns 1000 results. @tsenart's idea of us doing the filtering and just listing everything may be necessary, but also makes us lose some power / hard to do against public codehosts due to the amount of repos.

I am not sure about other APIs + search, but I don't think they are as limited as githubs w.r.t. rate limits.

**Comment by nicksnyder** [comment]
> github's search API has a rate limit of 20 requests an hour
> Additionally a search request only returns 1000 results

Are these limitations only on GitHub.com or do they also impact GitHub Enterprise?

**Comment by sqs** [comment]
GHE does not rate limit by default, but site admins can enable it: https://help.github.com/enterprise/2.16/admin/guides/installation/configuring-rate-limits/.

**Comment by nicksnyder** [comment]
Given FOSDEM and the fact that both Keegan and Tomás are working on 3.0 release blockers, this tracking issue will not get updated with a plan until next week.

**Comment by nicksnyder** [comment]
This is a priority to make progress on, so even if there is nothing to ship in 3.1, there should be a goal to ship some improvements here for 3.2.

**Comment by tsenart** [comment]
> This is a priority to make progress on, so even if there is nothing to ship in 3.1, there should be a goal to ship some improvements here for 3.2.

Sounds good to me.

**Comment by keegancsmith** [comment]
FYI we are actively working on this. Our current plan is to take the work @tsenart did and make it shippable just for github code hosts. Then based on what we learn, create a more detailed plan.

**Comment by keegancsmith** [comment]
### Update

We need to be concerned about the `external_*` columns on the repo table. They
can be unset, or not be unique as a triple. However, we need them to always be
set and unique. So we need to implement a migration process which is not just
a SQL migration, since we will need to fetch data from the upstream codehost
when the fields are not set. We must also identify which external service kind
a row belongs to, before attempting to contact a codehost. This functionality
probably exists in RepoLookup.

As a path forward, we should investigate making the uniqueness constraint only
apply to non-null columns. This way we can continue on our current path.

Side-note: We talked about using new columns instead, but those would also
require back-filling the values.

Github syncing is working and it can co-exist with the older syncers. It
handles deletes gracefully.

### Next
- Test renames.
- Make it more robust under failures.
- Tests.
- TODOs in code.

**Comment by keegancsmith** [comment]
Raw notes from today:

``` org-mode
- [ ] still support Kind in api call
- [ ] discuss URN and associating with an external service rather than ID
- [ ] figure out deduplication between sources that yield the same repos
- [ ] are external services soft deleted
- [ ] need to fix external service being deleted => delete repos
- [ ] to dedup repo metadata conflicts, just pick newest updated_at timestamp.
- [ ] ensure external_service_ids is the set of sources listing the repo
- [ ] id bigserial

Data model wishlist:
- Associate a repo with external services. 1+
- deduplication of repo rows that represent the same upstream repo.
- Make it easier to have more attributes for a repository without updating the table schema. (star count, is archive, etc)
- alias => Prevent conflicts on name in repo table.
- renames => do upserts enforcing uniqueness constraint on external id.
- id bigserial
- enforce that external_service_ids is a set

blacklist repos

- Directly visit => 
- Search for repo by name (just needs to match substring); Find results in repo => have a toggle just for this
- List repos in admin view
- Visit on codehost and extension communicates

1. sourcegraph/sourcegraph -> sourcegraph/enterprise
2. Created new repo sourcegraph/sourcegraph

1. sourcegraph/sourcegraph id=a
2. sourcegraph/enterprise id=a
3. sourcegraph/enterprise id=a; sourcegraph/sourcegraph id=b

upsert WHERE service_id = 


source_1 creating a row
source_2 creating a row   same

source_1 with different naming pattern
source_2 with default naming pattern    insert same upstream repo, but different names.


- We want name to be unique and canonical. Visiting sourcegraph.com/{name}
  needs to take you to the (only) repository and should map to the upstream
  code host.
- We want a repo to be eventually consistent with the upstream codehost:
  - name
  - metadata
  - existance
  - still configured to sync
- We want to support the OTHER codehost kind. OTHER is a code host without
  metadata (only name and clone URL). Need to handle conflicts with codehosts.
- We want to support multiple configurations for an external service
  kind. Example: Different search queries or different tokens.
```

**Comment by tsenart** [comment]
## Some notes from today's investigation.

### Global diff scalability

Currently loading all of sourcegraph.com's  ~31M repos from Postgres takes about 10s on my machine. This means that a sync interval of 10s in repo updater wouldn't make sense, but also, I was pleasantly surprised by how fast it is.

However, this number will only grow unbounded for sourcegraph.com in particular. It may not be something we need to address now, but eventually, we could need to find a way to create a diff between sourced repos (from external services) and stored repos (in the repo table) without having to load all stored repos in memory at once.

### `external_xxx` columns

The three existing columns `external_id`, `external_service_id` and `external_service_type` suit our needs of deduplicating sourced repos per code host instance. Since we can have multiple external services configured with the same code host instance, and the IDs of repos returned by those instances are guaranteed to be unique only in scope of those instances, we need to rely on a unique constraint on `(external_id, external_service_id)` for repository deduplication. Please note that the value of the `external_service_id` column is the URL of the code host.

```sql
select codehosts.domain, count(*)
from (select split_part(repo.name, '/', 1) as domain from repo) as codehosts
group by codehosts.domain;
                                   
+--------------------------+---------+
| domain                   | count   |
|--------------------------+---------|
| android.googlesource.com | 1       |
| bitbucket.org            | 7       |
| git.nssm.cc              | 1       |
| github.com               | 3141537 |
| gitlab.com               | 927     |
| hg.openjdk.java.net      | 2       |
+--------------------------+---------+
```

Unfortunately, it seems that many repo row don't have these columns filled in. This happens in situations when we [get rate limited](https://github.com/sourcegraph/sourcegraph/blob/master/cmd/repo-updater/repos/github.go#L217).

Here's the query to show that.

```sql
 select codehosts.domain, count(*) 
 from (
    select split_part(repo.name, '/', 1) as domain from repo
    where external_id is null and external_service_id is null
 ) as codehosts 
 group by codehosts.domain;       
                                                                                                                                                                                                                                                                                                                                                   
+--------------------------+---------+
| domain                   | count   |
|--------------------------+---------|
| android.googlesource.com | 1       |
| bitbucket.org            | 7       |
| git.nssm.cc              | 1       |
| github.com               | 749858  |
| hg.openjdk.java.net      | 2       |
+--------------------------+---------+
```

**Comment by nicksnyder** [comment]
> However, this number will only grow unbounded for sourcegraph.com in particular.

Historically, sourcegraph.com has been a special snowflake with respect to repo syncing (i.e. everything is on-demand instead of scheduled). Is that going to change in the future?

**Comment by tsenart** [comment]
> Is that going to change in the future?

Nope, good point!

**Comment by keegancsmith** [comment]
@tsenart an agenda for our discussion today:

- Should we consider feature flagging the new syncer so we can merge it into master before github support is polished?
- Discuss removing enabled/disabled + proposed migration strategy.
- Review kick-off document.

**Comment by keegancsmith** [comment]
> Should we consider feature flagging the new syncer so we can merge it into master before github support is polished?

Yes, this will be a goal for our work tomorrow. Once done we will review our code to ensure it is safe to merge into master.


> Discuss removing enabled/disabled + proposed migration strategy.

Yes we will do this. **This is a blocker for 3.2**. This is a proposed migration
path from our discussions, but not necessarily what we will do: A one-time,
best-effort process after upgrading to 3.2 will attempt to associate
repositories that aren't associated with an external service with a
best-effort related external service: all "github.com/" repositories will be
associated with the 1st GitHub external service (by adding them explicitly in
the "repos" array config value). That is the only heuristic needed. All
repositories not associated with an external service will be added to a new
"Other" external service named "Other repositories (pre-Sourcegraph 3.2
upgrade)" with their name and clone URL manually specified. The config JSON
for that external service will be:

```jsonc
// These repositories were not associated with a code host defined as an external service.
// To complete the migration to Sourcegraph 3.2 and ensure their metadata is updated,
// add them to the JSON configuration for their code host (or create a new code host external service if none exists).
{
   //...
}
```

We will only have Github migrated for 3.2. The above explanation is for
everyone. So we need to adapt the process to only work for github. The
migration process will only run on repos with an empty source column.

Our goal is to not remove repositories that are managed by the existing
configuration, since the pre-existing code would never remove/rename a
repository.


> Review kick-off document.

Done.

**Comment by tsenart** [comment]
@sqs: Product question. Do we still want admins to be able to delete a repo directly from the repo settings page? I feel this is slightly at odds with the configuration driven approach we're moving towards. Similarly to how the enabled / disabled state doesn't fit in anymore.

**Comment by sqs** [comment]
No. GitHub repositories (and in the future other repositories that are kept in sync with config) would not have delete buttons.

**Comment by keegancsmith** [comment]
@sqs do any of our current (or future) customers have a use case of some external services being managed by polling, but other external services being managed lazily. polling == our new syncer, what this issue is about. lazily == Sourcegraph.com behaviour of "just in time" syncing for a specific repo.

**Comment by sqs** [comment]
@keegancsmith Beyang is checking with the one customer that may currently rely on that behavior (to avoid overloading their Git host).

**Comment by beyang** [comment]
@keegancsmith we currently have one customer with the following situation:
* They have 11k+ repositories.
* `disableAutoGitUpdates` is true in their site config.
* They have configured an external script to hit the [repo refresh endpoint](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/httpapi/repo_refresh.go#L13:6-13:22) to manually trigger updates of repository contents.
* They have two code hosts: one is GitHub.com, the other is a private Gitolite instance.

Their goal is to ensure their repository contents are always up to date (including appearing in search results). They configured the manual update trigger to ensure this, because the polling behavior was insufficient as of a few months back (users were noticing repos out of date). An additional consideration is that their Gitolite API has rate limits, which the polling repo-updater also ran into earlier.

So they don't want actual "lazy" behavior, but I guess from repo-updater's perspective, they've configured "lazy" behavior by setting `disableAutoGitUpdates: true` and hitting the repo-refresh endpoint manually.

**Comment by nicksnyder** [comment]
It is important to disambiguate between refreshing repo metadata and refreshing the git contents.

Sourcegraph.com is lazy with respect to both metadata and git contents.

I can't imagine a scenario where a customer would want to be lazy with metadata, but as Beyang mentioned, we do have customers who want to have absolute control over fetching git contents so they can achieve optimal behavior (e.g. sourcegraph only updates when a commit is pushed).

**Comment by keegancsmith** [comment]
@beyang my question above was about targeted to metadata. I suppose in the case of gitolite there isn't much metadata to manage. The only real thing to manage is detection of deletion. In the case of the customer you mentioned, is it safe for us to poll the gitolite "list" repos endpoint? By default we will poll every minute (but is configurable)

**Comment by keegancsmith** [comment]
I have just updated the milestone and description to reflect this will ship in 3.3. We are focussed on shipping a very reliable service here, and given that branch cut is this Friday we do not feel comfortable that the state of the project by Friday will meet that quality bar. However, our goal for 3.3 does not change. So we have slipped a bit on shipping GitHub support, but we will still ship support for all code hosts in 3.3.

The feature is currently enabled on our dogfood instance behind a feature flag. We may be at a point by Friday where we feel comfortable suggesting to some customer to enable the feature flag. If that is the case by Friday I will update the description for instructions on how to try it out.

**Comment by beyang** [comment]
> @beyang my question above was about targeted to metadata. I suppose in the case of gitolite there isn't much metadata to manage. The only real thing to manage is detection of deletion. In the case of the customer you mentioned, is it safe for us to poll the gitolite "list" repos endpoint? By default we will poll every minute (but is configurable)

The customer currently has `disableAutoGitUpdates` set to true and relies on hitting the [repo refresh endpoint](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/httpapi/repo_refresh.go#L13:6-13:22) to trigger updates. If that's *not* going away, then we should be fine. If it is, then that will be an issue as the customer estimates load on their Gitolite API increases by 20-25% when polling is enabled and that puts us slightly over the load limit.

**Comment by keegancsmith** [comment]
@beyang The repo refresh endpoint is not going away. It is a different concern. To give an idea of how things fit together

* **Syncer** Responsible for communicating with the code host API to manage our set of repositories. It updates our repo table and informs **Scheduler** about our set of repositories.
* **Scheduler** Decides when to run `git fetch` or `git clone` for a repository.

We are only concerned about the syncer here, we will not be changing the behaviour of the scheduler. In this case what the syncer will run every minute is [ListGitolite](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8f3ebfb453ac6c77f87b53c6dbb79b6c0901bc5c/-/blob/cmd/repo-updater/repos/gitolite.go#L110) which just runs [ssh $HOST info](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8f3ebfb/-/blob/cmd/gitserver/server/list.go#L127).

To me this counts as us accessing the Gitolite API. However, it seems pretty lightweight. So I just want to confirm if that is alright.

**Comment by beyang** [comment]
Is this [an accurate representation of what the new arch will look like](https://drive.google.com/file/d/1Y0gEdbHyXkgKSUvUfRAAOPtzoVwwVgT7/view?usp=sharing) (you can edit this directly)? 

The once a minute sync should be fine (I believe it does that now).

**Comment by tsenart** [comment]
@beyang: I made these two changes to the diagram:

1. Removed dependency between scheduler and repos table.
1. Removed "update-repo endpoint" from Scheduler (not sure what you meant here)

In general, this diagram captures the system architecture as it was and as it is, since it hasn't changed. The larger changes were inside the repo-updater code architecture itself. Future system level architecture changes aren't planned as of yet.

**Comment by beyang** [comment]
So the scheduler doesn't directly read from the DB? It only gets triggered by the syncer? If it goes down, we may lose some updates (fine if this is the case; just clarifying my understanding).

**Comment by tsenart** [comment]
@beyang: The scheduler is in the same process as the syncer, all in repo-updater. When repo-updater goes down, the schedule is lost. While this hasn't changed with our recent work, we have ideas to improve the situation in the future by moving the scheduler logic to git-server and using a unified log of sorts for reliable at-least once delivery of events between repo-updater and git-server.

**Comment by nicksnyder** [comment]
My mental model: If repo-updater restarts then it reloads info from the db, queries gitserver for each known repo to be able to compute then next scheduled update (I.e. recreate the schedule). I don’t think anything would get lost here.

**Comment by tsenart** [comment]
> I don’t think anything would get lost here.

Nothing gets permanently lost, because, like you said, the schedule is recomputed eventually. But the original schedule is lost if repo-updater goes down.

**Comment by tsenart** [comment]
@sqs: Once we start implementing this work for other codehosts than GitHub, what how would you prioritise them?

**Comment by sqs** [comment]
@tsenart: GitHub, GitLab, Bitbucket Server, Phabricator (we don't sync Phab repos in the same way as we do for other code hosts, so there might not be anything needed for Phabricator; this is not implying you need to add repo syncing support), Gitolite, other (Git URLs), AWS CodeCommit

**Comment by tsenart** [comment]
Hey @francisschmaltz,

After our retro, I thought it would be good to get your input on the above UI/UX related aspects of this project. What's the best way to bring you on-board?

**Comment by francisschmaltz** [comment]
> @tsenart What's the best way to bring you on-board?

The best way is to set up a meeting where we can go over the requirements for the UI and then talk through and maybe wireframe some ideas and I can schedule time for mocks to be made.

**Comment by tsenart** [comment]
I tagged you in the relevant checklist item in the issue description, @francisschmaltz.

**Comment by tsenart** [comment]
> The best way is to set up a meeting where we can go over the requirements for the UI and then talk through and maybe wireframe some ideas and I can schedule time for mocks to be made.

Sounds good. How about next Monday after team meeting?

**Comment by keegancsmith** [comment]
I read over the remaining work. I think we should prioritize implementing all
other code hosts. From what I can tell other code hosts pose the biggest
unknown risk. For example just for GitHub we have noticed two blocking issues
(repositoryQuery with more than 1k results, non-existant repos in
github.repos). Additionally once every code host is supported, the mental
model of owning the whole table (instead of partially) will likely simplify a
few things.

**Comment by tsenart** [comment]
I have cleaned up and update the this tracking issue. We'll be talking about it in more detail during team meeting.

@francisschmaltz: The UI work is marked clearly at the top. Once we have your designs we can proceed. Please note that the UI work items of each external service's work items are dependent on the top-level UI work.

**Comment by francisschmaltz** [comment]
Design links added to the bottom of the issue description

> Design
Sourcegraph Engineer / Design Spec Link:
https://zpl.io/anpXvGr
Public / Non-Engineer Design Link:
https://sketch.cloud/s/wwoGV/Axae1bR/play

**Comment by tsenart** [comment]
@ryan-blunden: Answering your questions here instead of the issue description:

- When a repository is no longer in the computed list, are its contents deleted from disk?
  - Answer: When a repository is deleted, its contents (git data) will eventually be deleted from disk.
- Is there a UI to view the computed list of repositories?
  - Answer: The existing repository listing page will serve this purpose. See [design mocks](https://github.com/sourcegraph/sourcegraph/issues/2025#issuecomment-479232898).
- Is there a log of operations including when configuration values change?
  - Answer: We don't have an audit-log feature at this time. Our operational logs will output detail information of changes made when in DEBUG level.
- What's the impact on code hosts that do not have this feature yet, e.g. AWS CodeCommit or Gitolite?
  - Answer: No impact, although there are small UI changes in repo settings page as you can see in the [design mocks](https://github.com/sourcegraph/sourcegraph/issues/2025#issuecomment-479232898).

**Comment by tsenart** [comment]
@keegancsmith: Tomorrow let's hold a ceremony to close off this issue :-)

**Comment by francisschmaltz** [comment]
@tsenart @keegancsmith before we close can we open a tracking issue for getting the UI inline with mocks for 3.4

**Comment by keegancsmith** [comment]
> @tsenart @keegancsmith before we close can we open a tracking issue for getting the UI inline with mocks for 3.4

https://github.com/sourcegraph/sourcegraph/issues/3472

**Comment by slimsag** [comment]
I'll leave open for your closing ceremony :) Amazing work here.

Be sure to close / move other stuff to next milestone if important.

**Comment by tsenart** [comment]
:shipit:
